js/web/lib/wasm/jsep/webgpu/ops/gather-elements.ts

```diff
@@ -32,65 +32,68 @@ const createGatherElementsProgramInfo =
-      const indicesSize = ShapeUtil.size(indicesShape);
-
       const axis = ShapeUtil.normalizeAxis(attributes.axis, inputRank);
       const axisDimLimit = inputShape[axis];
 
       const outputShape = indicesShape.slice(0);
       const outputSize = ShapeUtil.size(outputShape);
+      const enableInputShapesUniforms = enableShapesUniforms(inputShape.length);
+      const inputShapeOrRank = enableInputShapesUniforms ? inputShape.length : inputShape;
+      const enableIndicesOutputShapesUniforms = enableShapesUniforms(indicesShape.length);
+      const indicesOrOutputShapeOrRank = enableIndicesOutputShapesUniforms ? indicesShape.length : indicesShape;
 
-      const input = inputVariable('input', inputOutputDataType, inputShape);
-      const indices = inputVariable('indices', indicesDataType, [indicesSize]);
-      const output = outputVariable('output', inputOutputDataType, outputShape);
+      const input = inputVariable('input', inputOutputDataType, inputShapeOrRank);
+      const indices = inputVariable('indicesInput', indicesDataType, indicesOrOutputShapeOrRank);
+      const output = outputVariable('output', inputOutputDataType, indicesOrOutputShapeOrRank);
 
 
+      const programUniforms: ProgramUniform[] =
+          [{type: 'uint32', data: outputSize}, {type: 'int32', data: axisDimLimit}, {type: 'uint32', data: axis}];
+      if (enableInputShapesUniforms) {
+        programUniforms.push(...createTensorShapeVariables(inputShape));
+      }
+      if (enableIndicesOutputShapesUniforms) {
+        programUniforms.push(...createTensorShapeVariables(indicesShape));
+        programUniforms.push(...createTensorShapeVariables(outputShape));
+      }
+      const inputDependencies: ProgramInputTensorInfoDependency[] =
+          [enableInputShapesUniforms ? 'rank' : 'dims', enableIndicesOutputShapesUniforms ? 'rank' : 'dims'];
+
       // int64 indices would be treated as little endian i32 with assumption they fall in i32 limits
       // That assumption is safe as it's not possible to allocate >2gb buffer for input tensor
       // Input data will be treated as u32 or two u32 for 8-byte tensors
       const getShaderSource = (shaderHelper: ShaderHelper) => `
-      const inputStrides = array<u32, ${inputStrides.length}>(${inputStrides.map(i => `${i}u`).join(',')});
-      ${shaderHelper.declareVariables(input, indices, output)}
+      ${
+          shaderHelper.registerUniform('outputSize', 'u32')
+              .registerUniform('axisDimLimit', 'i32')
+              .registerUniform('axis', 'u32')
+              .declareVariables(input, indices, output)}
       ${shaderHelper.mainStart()}
-      ${shaderHelper.guardAgainstOutOfBoundsWorkgroupSizes(outputSize)}
+      ${shaderHelper.guardAgainstOutOfBoundsWorkgroupSizes('uniforms.outputSize')}
 
       let outputIndices = ${output.offsetToIndices('global_idx')};
 
-      var idx = ${indices.getByOffset('global_idx')};
+      var idx = ${indices.getByIndices('outputIndices')};
```

Comment:
I think `getByoffset` is faster than `getByIndices`. Is there any special reason to change it to `getByIndices`? Maybe you can add a comment that output has the same shape as indices so that we can directly use `indices.getByOffset`.